contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,826 @@
+
+ 	// CurrentRevision, if not empty, indicates the version of PodSpecTemplate, 
+ 	// VolumeClaimsTemplate tuple used to generate Pods in the sequence
+ 	// [0,CurrentReplicas).
+ 	CurrentRevision string `json:"currentRevision,omitempty"`
+ 	
+ 	// UpdateRevision, if not empty, indicates the version of PodSpecTemplate, 
+    // VolumeClaimsTemplate tuple used to generate Pods in the sequence
+    // [Replicas-UpdatedReplicas,Replicas)
+ 	UpdateRevision string `json:"updateRevision,omitempty"`
+ 	
+ 	// ReadyReplicas is the current number of Pods, created by the StatefulSet
+    // controller, that have a Status of Running and a Ready Condition.
+ 	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+ 	
+ 	// CurrentReplicas is the number of Pods created by the StatefulSet 
+    // controller from the PodTemplateSpec, VolumeClaimsTemplate tuple indicated 
+    // by CurrentRevision.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+ 	
+ 	// UpdatedReplicas is the number of Pods created by the StatefulSet
+    // controller from the PodTemplateSpec, VolumeClaimsTemplate tuple indicated 
+    // by UpdateRevision.
+ 	UpdatedReplicas int32 `json:"taretReplicas,omitempty"`
+}
+```
+
+Additionally we introduce the following constant.
+
+```go
+// StatefulSetRevisionLabel is the label used by StatefulSet controller to track
+// which version of StatefulSet's StatefulSetSpec was used generate a Pod.
+const StatefulSetRevisionLabel = "statefulset.kubernetes.io/revision"
+
+```
+## StatefulSet Controller
+The StatefulSet controller will watch for modifications to StatefulSet and Pod 
+API objects. When a StatefulSet is created or updated, or when one 
+of the Pods in a StatefulSet is updated or deleted, the StatefulSet
+controller will attempt to create, update, or delete Pods to conform the 
+current state of the system to the user declared [target state](#target-state). 
+
+### Revised Controller Algorithm
+The StatefulSet controller will use the following algorithm to continue to 
+make progress toward the user declared [target state](#target-state) while 
+respecting the controller's 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+[deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees. The StatefulSet controller will use the technique proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594) to 
+snapshot and version its [target Object state](#target-pod-state).
+
+1. The controller will reconstruct the 
+[revision history](#history-reconstruction) of the StatefulSet.
+1. The controller will 
+[process any updates to its StatefulSetSpec](#specification-updates) to 
+ensure that the StatefulSet's revision history is consistent with the user 
+declared desired state.
+1. The controller will select all Pods in the StatefulSet, filter any Pods not 
```

Comment:
Controller overlap is possible if multiple controllers are created with overalpping selectors. We manage this with ControllerRef's for a ll controllers as of 1.6.